use GitHub Actions to build and publish bottles

dpo opened the pull request:
The "test" workflow will trigger builds of formulae updated in pull requests and upload bottles (i.e., precompiled binaries) as artifacts. The "publish" workflow will publish those artifacts as releases in this repository and update the formulae with a "bottle" section.

I've been testing this this week. It's not bulletproof in the sense that the "test" workflow is sometimes shown as having succeeded, yet bottles were not produced because of an error in one of the  steps. So it requires a bit of manual inspection. Still it's a huge help.

In order to publish bottles, a repository owner must label the pull request with the "pr-pull" label (I just created it). If all goes well, the bottles will be published as releases and the pull request will be closed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/coin-or-tools/homebrew-coinor?pullRequest=68) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/coin-or-tools/homebrew-coinor?pullRequest=68) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/coin-or-tools/homebrew-coinor?pullRequest=68) it.</sub>

dpo commented:
Brew is very unhappy about stylistic issues here.

dpo commented:
I temporarily disabled the stylistic checks, but those should be re-enabled eventually. They've evolved quite a bit in Homebrew (and are a moving target) since we've created this tap.

Among others, if you're interested in this CI workflow, we'll have to revisit each formula.

tkralphs commented:
This is great, thanks! Any idea about the circleci failure? I've seen that one before. It doesn't seem related to the recipes.

tkralphs commented:
Currently, on Linuxbrew, I get the following.
```
~/BTSync/dotfiles > brew info cbc
cbc: stable 2.10.6 (bottled)
Mixed integer linear programming solver
https://github.com/coin-or/Cbc
/home/ted/.linuxbrew/Cellar/cbc/2.10.2_1 (97 files, 3.0MB) *
  Built from source on 2019-11-16 at 22:35:14
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/cbc.rb
License: EPL-1.0
==> Dependencies
Build: pkg-config ✘
Required: cgl ✘, clp ✘, coinutils ✘, osi ✘
==> Analytics
install: 447 (30 days), 922 (90 days), 2,236 (365 days)
install-on-request: 251 (30 days), 516 (90 days), 1,272 (365 days)
build-error: 0 (30 days)
```
It says I have 2.10.6 installed and when I brew edit, the recipe is for 2.10.6, but the location where it's supposed to be installed is `/home/ted/.linuxbrew/Cellar/cbc/2.10.2_1` and when I run Cbc, I get version 2.10.2, not 2.10.6.

I noticed in the output for the CI this line: https://github.com/coin-or-tools/homebrew-coinor/runs/4903828061?check_suite_focus=true#step:8:6

which says nothing was uploaded. I'm a little confused what is going on here.

dpo commented:
The reason for the confusion may be that Homebrew Core already contains some CoinOR formulae. The output of your `brew info cbc` says:
```
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/cbc.rb
```
Compare with
```
$ brew info coin-or-tools/coinor/cbc
coin-or-tools/coinor/cbc: stable 2.10.3, HEAD
Mixed integer linear programming solver
https://projects.coin-or.org/Cbc
Not installed
From: https://github.com/coin-or-tools/homebrew-coinor/blob/HEAD/cbc.rb
==> Dependencies
Build: pkg-config ✔
Required: gcc ✔, coin-or-tools/coinor/cgl ✘, coin-or-tools/coinor/clp ✘
Optional: ampl-mp ✔, coin-or-tools/coinor/glpk448 ✔, dpo/openblas/mumps --without-mpi ✘, openblas ✔, suite-sparse ✘
==> Options
--with-ampl-mp
        Build with ampl-mp support
--with-glpk448
        Build with glpk448 support
--with-mumps
        Build with mumps support
--with-openblas
        Build with openblas support
--with-parallel
        Build with parallel mode enabled
--with-suite-sparse
        Build with suite-sparse support
--HEAD
        Install HEAD version
==> Analytics
install: 447 (30 days), 922 (90 days), 2,236 (365 days)
install-on-request: 251 (30 days), 516 (90 days), 1,272 (365 days)
build-error: 0 (30 days)
```

dpo commented:
> I noticed in the output for the CI this line: https://github.com/coin-or-tools/homebrew-coinor/runs/4903828061?check_suite_focus=true#step:8:6
>
> which says nothing was uploaded. I'm a little confused what is going on here.

Yes, this PR only introduces the workflow files. Bottles are only built for PRs that modify one or more formulae (this PR doesn't).

My objective here is
1) to show you what's possible with GitHub Actions;
2) to illustrate the fact that most/all formulae will have to be revised before they pass the syntax checks;
3) that in particular, we'll have to decide on sets of default options because Homebrew no longer wants formulae that pass options down to other formulae.

tkralphs commented:
> The reason for the confusion may be that Homebrew Core already contains some CoinOR formulae. 

Ah, you are right, I forgot about that! But I guess this is then a bug in homebrew then.

dpo commented:
@tkralphs I submitted updates to all formulae. They're building now. Finger crossed.

Updates:
* I removed all options and activated all dependencies whenever possible;
* there is a new formula for sequential MUMPS;
* I renamed glpk448 to glpk@448 (Homebrew's syntax for versioned formulae).

Things to think about:
* activating the SuiteSparse dependency in CLP results in a build error (I left it as `:optional` for now);
* homebrew/core already contains CoinUtils, CGL, CLP, CBC and OSI but not always with the same set of dependencies as here;
* I declared that formulae that already exist in core should be "keg only" (i.e., not installed to `/usr/local`);
* `vol.rb` says that the dependency on `osi` is `:recommended`, but I don't see anything about OSI in the `configure` command. Is it autodetected? Shouldn't it be activated all the time (i.e., remove the `:recommended`)?

Let's know what you think of the changes.

dpo commented:
- [ ] There's an error on Linux (unfortunately, it's not causing the CI job to fail):
```
==> brew linkage --test coin-or-tools/coinor/coinutils
==> FAILED
Full linkage --test coin-or-tools/coinor/coinutils output
  Unwanted system libraries:
    /lib/x86_64-linux-gnu/libbz2.so.1.0
```
Why does `coinutils` link against `libbz2`? I don't see any option related to it in `configure --help`.

tkralphs commented:
> * [ ]  There's an error on Linux (unfortunately, it's not causing the CI job to fail):
> 
> ```
> ==> brew linkage --test coin-or-tools/coinor/coinutils
> ==> FAILED
> Full linkage --test coin-or-tools/coinor/coinutils output
>   Unwanted system libraries:
>     /lib/x86_64-linux-gnu/libbz2.so.1.0
> ```
> 
> Why does `coinutils` link against `libbz2`? I don't see any option related to it in `configure --help`.

Yes, there is a check for libbz2 and it does link to it, if found. It's used to read compressed input files. You probably checked he help on the wrong configure scripts. It's not the root script, it's the one in the CoinUtils subdirectory that actually matters.

tkralphs commented:
> Things to think about:
> 
> * activating the SuiteSparse dependency in CLP results in a build error (I left it as `:optional` for now);

I'll try to reproduce in Linux. 

> * homebrew/core already contains CoinUtils, CGL, CLP, CBC and OSI but not always with the same set of dependencies as here;

Yes, I'm aware, but as far as I remember, they did not include all the options that I would have and were a bit inflexible to deal with. If we had our own versions that we could control, as long as it's not a huge amount of work, I wouldn't mind. But we can discuss. 

> * I declared that formulae that already exist in core should be "keg only" (i.e., not installed to `/usr/local`);

I'm not sure I see the implications of this. If someone explicitly asks to install a package from this tap, I would think it should be installed. 

> * `vol.rb` says that the dependency on `osi` is `:recommended`, but I don't see anything about OSI in the `configure` command. Is it autodetected? Shouldn't it be activated all the time (i.e., remove the `:recommended`)?

Again, I think you're looking at the wrong configure script. There is an OsiVol that will be built if Osi is present.

tkralphs commented:
From what I understand, "keg only" will mean that if you install cbc from this tap, it's just not going to work, even if you don't have the one from core installed. In other words, you won't have the `cbc` executable in your path and won't be able to link to `libCbc`. On the other hand, the downside of not marking them keg only seems small. It's possible someone would install both, but this won't actually cause a problem as long as the versions line up. For now, I'm leaning towards not doing keg only and then if we get bug reports, we can deal with it. But otherwise, I need to change the instructions in the Cbc README to steer people away from installing Cbc using these formulae because they won't work. Am I understanding correctly?

dpo commented:
The main issue is that every Homebrew user will have the "core" tap. It's ok for formulae to be keg only. You just have to add their location to your PATH and LD_LIBRARY_PATH. Brew tells you about it when you install a formula. You can also get that info later with
```
$ brew info coin-or-tools/coinor/clp
coin-or-tools/coinor/clp: stable 1.17.6, HEAD [keg-only]
Linear programming solver
https://github.com/coin-or/Clp
/usr/local/Cellar/clp/1.17.6_1 (74 files, 2.9MB)
  Built from source on 2022-01-31 at 20:05:22
From: https://github.com/coin-or-tools/homebrew-coinor/blob/HEAD/clp.rb
==> Dependencies
Build: pkg-config ✔
Required: ampl-mp ✔, coin-or-tools/coinor/coinutils ✔, coin-or-tools/coinor/glpk@448 ✔, coin-or-tools/coinor/mumps-seq ✔, coin-or-tools/coinor/osi ✔, gcc ✔, readline ✔
Recommended: openblas ✔
Optional: suite-sparse ✔
==> Options
--with-suite-sparse
        Build with suite-sparse support
--without-openblas
        Build without openblas support
--HEAD
        Install HEAD version
==> Caveats
clp is keg-only, which means it was not symlinked into /usr/local,
because conflicts with formula in core.

If you need to have clp first in your PATH, run:
  echo 'export PATH="/usr/local/opt/clp/bin:$PATH"' >> /Users/dpo/.bash_profile

For compilers to find clp you may need to set:
  export LDFLAGS="-L/usr/local/opt/clp/lib"
  export CPPFLAGS="-I/usr/local/opt/clp/include"

For pkg-config to find clp you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/clp/lib/pkgconfig"

==> Analytics
install: 618 (30 days), 1,095 (90 days), 2,638 (365 days)
install-on-request: 112 (30 days), 168 (90 days), 418 (365 days)
build-error: 0 (30 days)
```

If you really want/need to have it on your default search path you can use `brew link --force`:
```
$ brew link clp
Warning: clp is keg-only and must be linked with `--force`.

If you need to have this software first in your PATH instead consider running:
  echo 'export PATH="/usr/local/opt/clp/bin:$PATH"' >> /Users/dpo/.bash_profile
```

The best would be to convince the core team to remove COIN packages from the core tap, but I'm not sure that's going to happen.

dpo commented:
@tkralphs The most recent build succeeded. Have a look through the most recent run of the GitHub Action. It's happy on Ubuntu, macOS Big Sur and macOS Catalina. The artifacts zip file contains 104MB of precompiled packages.

Also have a look through the updated formulae to see if you agree with the changes.

I left `openblas => :recommended` everywhere so you can build from source `--without-openblas` (and cross fingers that `configure` autodetects your BLAS and LAPACK).

Let me know what you think. Once we're happy and all the bottles build, all it takes is to add the "pr-pull" label to this PR and everything else should automagically happen: the precompiled bottles will be registered as releases on this repo, the formulae will be updated with the bottle download info, committed to the master branch, and this PR will be closed.

tkralphs commented on dylp.rb:
```diff
@@ -3,6 +3,8 @@ class Dylp < Formula
   homepage "https://projects.coin-or.org/DyLP"
```
This is https://github.com/coin-or/DyLP (all projects are on Github now.

tkralphs commented on osi.rb:
```diff
@@ -3,32 +3,41 @@ class Osi < Formula
 
-  glpk_dep = build.with?("glpk") ? ["with-glpk"] : []
-  openblas_dep = build.with?("openblas") ? ["with-openblas"] : []
+  head "https://projects.coin-or.org/svn/Osi/trunk"
```
Same.

tkralphs commented on symphony.rb:
```diff
@@ -3,28 +3,18 @@ class Symphony < Formula
   homepage "https://projects.coin-or.org/SYMPHONY"
```
This should be https://github.com/coin-or/SYMPHONY

tkralphs commented on symphony.rb:
```diff
@@ -3,28 +3,18 @@ class Symphony < Formula
   homepage "https://projects.coin-or.org/SYMPHONY"
   url "https://www.coin-or.org/download/pkgsource/SYMPHONY/SYMPHONY-5.6.17.tgz"
```
https://github.com/coin-or/SYMPHONY/archive/refs/tags/releases/5.6.18.tar.gz

tkralphs commented on vol.rb:
```diff
@@ -3,6 +3,8 @@ class Vol < Formula
   homepage "https://projects.coin-or.org/Vol"
```
https://github.com/coin-or/Vol

tkralphs commented on vol.rb:
```diff
@@ -3,6 +3,8 @@ class Vol < Formula
   homepage "https://projects.coin-or.org/Vol"
   url "https://www.coin-or.org/download/pkgsource/Vol/Vol-1.5.4.tgz"
```
https://github.com/coin-or/Vol/archive/refs/tags/releases/1.5.4.tar.gz

tkralphs commented on dylp.rb:
```diff
@@ -3,6 +3,8 @@ class Dylp < Formula
   homepage "https://projects.coin-or.org/DyLP"
   url "https://www.coin-or.org/download/pkgsource/DyLP/DyLP-1.10.4.tgz"
```
https://github.com/coin-or/DyLP/archive/refs/tags/releases/1.10.4.tar.gz

tkralphs reviewed (changes requested):
Looks good, but found a bunch of references to projects.coin-or.org. Everything has been moved to Github.

tkralphs commented on coin_data_miplib3.rb:
```diff
@@ -3,6 +3,7 @@ class CoinDataMiplib3 < Formula
   homepage "https://www.coin-or.org/download/source/Data"
```
https://github.com/coin-or-tools/Data-miplib3

tkralphs commented on coin_data_miplib3.rb:
```diff
@@ -3,6 +3,7 @@ class CoinDataMiplib3 < Formula
   homepage "https://www.coin-or.org/download/source/Data"
   url "https://www.coin-or.org/download/source/Data/Data-miplib3-1.2.7.tgz"
```
https://github.com/coin-or-tools/Data-miplib3/archive/refs/tags/releases/1.2.8.tar.gz

tkralphs commented on coin_data_netlib.rb:
```diff
@@ -3,6 +3,7 @@ class CoinDataNetlib < Formula
   homepage "https://www.coin-or.org/download/source/Data"
```
https://github.com/coin-or-tools/Data-netlib

tkralphs commented on coin_data_netlib.rb:
```diff
@@ -3,6 +3,7 @@ class CoinDataNetlib < Formula
   homepage "https://www.coin-or.org/download/source/Data"
   url "https://www.coin-or.org/download/source/Data/Data-Netlib-1.2.8.tgz"
```
https://github.com/coin-or-tools/Data-Netlib/archive/refs/tags/releases/1.2.10.tar.gz

tkralphs commented on coin_data_sample.rb:
```diff
@@ -3,6 +3,7 @@ class CoinDataSample < Formula
   homepage "https://www.coin-or.org/download/source/Data"
```
https://github.com/coin-or-tools/Data-Sample

tkralphs commented on coin_data_sample.rb:
```diff
@@ -3,6 +3,7 @@ class CoinDataSample < Formula
   homepage "https://www.coin-or.org/download/source/Data"
   url "https://www.coin-or.org/download/source/Data/Data-Sample-1.2.11.tgz"
```
https://github.com/coin-or-tools/Data-Sample/archive/refs/tags/releases/1.2.12.tar.gz

tkralphs commented on coin_data_stochastic.rb:
```diff
@@ -3,6 +3,7 @@ class CoinDataStochastic < Formula
   homepage "https://www.coin-or.org/download/source/Data"
```
https://github.com/coin-or-tools/Data-stochastic

tkralphs commented on coin_data_stochastic.rb:
```diff
@@ -3,6 +3,7 @@ class CoinDataStochastic < Formula
   homepage "https://www.coin-or.org/download/source/Data"
   url "https://www.coin-or.org/download/source/Data/Data-Stochastic-1.1.6.tgz"
```
https://github.com/coin-or-tools/Data-Stochastic/archive/refs/tags/releases/1.1.7.tar.gz

tkralphs reviewed (changes requested):
Oops, I missed all the data projects, which are also on Github.

dpo commented:
Should all be fixed. 🤞

tkralphs commented:
@dpo Looks great, now! Thanks for doing all that. Can you just quickly sign the CLA?

dpo commented:
@tkralphs Done!

FYI I found that it's necessary to have a quick look at the detail of the Actions (for each architecture) to scan for hidden failures. Sometimes, one of the formulae will fail for some reason, and not make the whole job fail. I think that happens when a PR updates multiple formulae (another reason to avoid that).

In this case, I think it looks good. You have the honor to label this PR with the "pr-pull" label.

tkralphs commented:
I just discovered that Cgl fails to compile Vol is already installed, but we can deal with that later. As for having to look at the details of the Actions, that kinda defeats the purpose of CI :(. I would guess that this is configurable somehow. Perhaps whoever designed the action considered it a feature, not a bug, since it means the whole job won't be cancelled because of a single failure. But maybe we can look into that and customize if it becomes an issue over time. Thanks for pushing this over the finish line (although I guess we still need to update the versions)!

tkralphs commented:
I figured out the bug in Cgl. It's silly, but will need another release to fix that.

tkralphs commented:
Hmm, I gave it the label pr-pull and I see the action you added that's supposed to fire. I did see that all the tests seemed to re-run, but nothing else happened. Does the branch the PR is being merged into have to have Github Actions already deployed? I'm not sure if I should just go ahead and merge now or what. 

Also, why do bottles only get published for Ubuntu?

dpo commented:
I agree about the CI. I'll have to ask on the Homebrew forum. They ship those scripts.

Sorry about the bottles. I guess the CI config must be in `master` already before this can work. I'll create a new PR with just the CI scripts, merge it, and then rebase this one.

dpo commented:
Nice!

tkralphs commented:
OK, now it worked!
